models/buffer_disk.go

```diff
@@ -94,26 +94,26 @@ func (b *DiskBuffer) Add(metrics ...telegraf.Metric) int {
 	b.Lock()
 	defer b.Unlock()
 
-	dropped := 0
+	var batch wal.Batch
+	idx := b.writeIndex()
+	startIdx := idx
 	for _, m := range metrics {
-		if !b.addSingleMetric(m) {
-			dropped++
+		data, err := metric.ToBytes(m)
+		if err != nil {
+			panic(err)
 		}
+		batch.Write(idx, data)
+		idx++
 	}
-	b.BufferSize.Set(int64(b.length()))
-	return dropped
-}
 
-func (b *DiskBuffer) addSingleMetric(m telegraf.Metric) bool {
-	data, err := metric.ToBytes(m)
-	if err != nil {
-		panic(err)
-	}
-	if err := b.file.Write(b.writeIndex(), data); err != nil {
-		return false
+	if err := b.file.WriteBatch(&batch); err != nil {
+		dropped := uint64(len(metrics)) - (b.writeIndex() - startIdx)
```

Comment:
this method of finding how much of a batch was written assumes that there is only one concurrent writer to the WAL.